### PR TITLE
optimize: Add the server.WithBasePath function

### PR DESCRIPTION
**Pull request by springrain**
#### What type of PR is this?
optimize

#### Check the PR title.
- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.

#### (Optional) Translate the PR title into Chinese.
增加server.WithBasePath函数

#### (Optional) More detail description for this PR(en: English/zh: Chinese).
en: Similar to the java web's contextPath, it is convenient for nginx distribution
zh: 类似java web的contextPath,方便nginx分发

### CODE REVIEW

File: pkg/app/server/option.go [hunk 1]
```diff
@@ -151,6 +152,16 @@ func WithHostPorts(hp string) config.Option {
 	}}
 }
 
+// WithBasePath sets basePath
+func WithBasePath(basePath string) config.Option {
```
Comment (li-jin-gou) [inline pkg/app/server/option.go:156]: Unit tests need to be added.

**Review by li-jin-gou — CHANGES_REQUESTED** [review]

File: pkg/app/server/option.go [hunk 2]
```diff
@@ -151,6 +152,16 @@ func WithHostPorts(hp string) config.Option {
...
+	return config.Option{F: func(o *config.Options) {
+		if !strings.HasSuffix(basePath, "/") {
```
[the remark this answers is not in the document]
Comment (springrain) [reply pkg/app/server/option.go:162]: The annotations have been improved

File: pkg/app/server/option.go (hunk 1 above)
Comment (springrain) [reply pkg/app/server/option.go:156]: ok

**Comment by welkeyever** [comment]
lint needs to be fixed 🙏
Run `gofumpt -l -extra . ` to check.

File: pkg/app/server/option.go (hunk 1 above)
Comment (li-jin-gou) [reply pkg/app/server/option.go:156]: I think the code is OK, but I don't see a joint test with Hertz server in the unit test section, only a unit test for option. Similar to [it ](https://github.com/cloudwego/hertz/blob/develop/pkg/app/server/hertz_test.go#L307)

**Comment by welkeyever** [comment]
Please resign the CLA🙏

**Comment by springrain** [comment]
> lint needs to be fixed 🙏 Run `gofumpt -l -extra . ` to check.

I have already executed the command locally and there is no problem, I don't know why the check still fails.
Please help to modify the code format, thank you very much

**Comment by li-jin-gou** [comment]
> > lint needs to be fixed 🙏 Run `gofumpt -l -extra . ` to check.
> 
> I have already executed the command locally and there is no problem, I don't know why the check still fails. Please help to modify the code format, thank you very much

ok

**Comment by li-jin-gou** [comment]
> > lint needs to be fixed 🙏 Run `gofumpt -l -extra . ` to check.
> 
> I have already executed the command locally and there is no problem, I don't know why the check still fails. Please help to modify the code format, thank you very much

done and I use `gofumpt -e -d -w -extra .`

**Review by li-jin-gou — CHANGES_REQUESTED** [review]
The unit test did not pass and need to fix.

**Review by Duslia — APPROVED** [review]

**Review by li-jin-gou — APPROVED** [review]

**Comment by li-jin-gou** [comment]
LGTM

**Pull request merged**